HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc

```diff
@@ -811,9 +814,18 @@ void NVProfilerService::postModuleEventPrefetching(edm::StreamContext const& sc,
 }
 
 void NVProfilerService::preModuleConstruction(edm::ModuleDescription const& desc) {
+  auto mid = desc.id();
+  global_modules_.grow_to_at_least(mid + 1);
+
+  // This normally does nothing because stream_modules_ is empty when
+  // called. But there is a rare case when a looper is used that replacement
+  // modules can be constructed at end of loop. I'm not sure if that feature
+  // is ever actually used but just to be safe...
+  for (auto& modulesForOneStream : stream_modules_) {
```

Comment:
@wddgit note that `stream_modules_` is a regular `std::vector<std::vector<...>>`.

Can multiple `preModuleConstruction` happen in parallel in different threads ?

If so, and for any reason `stream_modules_` is not empty, this will try to resize the same `vector` concurrently, which is not thread safe.

Reply:
In the present version of the Framework, modules are constructed serially. There is not any concurrency. We've sometimes talked about changing that, but it would be difficult and would cause a lot of difficult to find and fix problems here and elsewhere.

stream_modules will always be empty in preModuleConstruction except in the case where it is called during end of loop via the looper. The end of loop code that replaces modules is serial.

And one other thing related that might be worth mentioning is that the source constructor is called concurrently with module construction normally. Sources are never constructed at end of loop so that is not an issue. Also, according to the TBB documentation grow_to_at_least can be called concurrently.

Reply:
Indeed, I'm not concerned about `global_modules_`: that is a `tbb::concurrent_vector` specifically to support growing in parallel, in `preSourceConstruction` and in `preModuleConstruction`.

The only concern I have is with the `modulesForOneStream.resize(...)`.

If the `preModuleConstruction` will always be called one module at a time, then it should be fine.